src/qibo/base/hamiltonians.py

```diff
@@ -187,6 +211,179 @@ def __matmul__(self, o):
                                              "implemented.".format(type(o)))
 
 
+class SymbolicHamiltonian:
+    """Parses symbolic Hamiltonians defined using ``sympy``.
+
+    This class should not be used by users.
+    It is used internally to help creating
+    :class:`qibo.base.hamiltonians.Hamiltonian` and
+    :class:`qibo.base.hamiltonians.TrotterHamiltonian` objects for Hamiltonians
+    defined using symbols. For more information we refer to the
+    :meth:`qibo.base.hamiltonians.Hamiltonian.from_symbolic`
+    and :meth:`qibo.base.hamiltonians.TrotterHamiltonian.from_symbolic` methods.
+
+    Args:
+        symbolic_hamiltonian: The full Hamiltonian written with symbols.
+        symbol_map (dict): Dictionary that maps each symbol to a pair of
+            (target, matrix).
+    """
+    import sympy
+    from qibo import matrices
+
+    def __init__(self, hamiltonian, symbol_map):
+        self.symbolic = self.sympy.expand(hamiltonian)
```

Comment:
Where is this `self.sympy` defined?
Shall we include checks for the hamiltonian and symbol_map types?

Reply:
> Where is this `self.sympy` defined?

`sympy` is imported within the `_SymbolicHamiltonian` class.

> Shall we include checks for the hamiltonian and symbol_map types?

I added some checks that the given arguments have the proper types and format. Let's wait for the coverage check too.